### ISSUE TITLE
Improve bitmap font rendering on macOS.

### ISSUE THREAD

**Issue by jeremija** [alacritty/alacritty#3459]
This is an example screenshot from iTerm:

![image](https://user-images.githubusercontent.com/1489493/76687851-18dbbb80-6628-11ea-8bec-c821d9fba842.png)

In iTerm's font configuration, the anti-aliased checkbox is not checked:

![image](https://user-images.githubusercontent.com/1489493/76687870-2f821280-6628-11ea-8ae6-a671861eabed.png)

Is there a way to turn off antialiasing and use the Monaco bitmap font in Alacritty on macOS?

Thanks!

### System

OS: macOS
Version: alacritty 0.4.1 (4d9253b)

**Comment by kchibisov** [comment]
we're not iterm.

**Comment by kchibisov** [comment]
Oh, maybe I don't get your issue, but on screenshot font is broken, since I can't even read it. If it's iterm. I think anti aliasing can't help it.

**Comment by kchibisov** [comment]
Oh, and AA related tweaks are a part of #957 , so you can't do that right now on macOS, but it should be possible after #957, so closing as a duplicate of it.

**Comment by jeremija** [comment]
@kchibisov I'm afraid the screenshot got a little messed up during upload (compression?) so it looks a little messed up, but I find it easier to read on a non-retina screen. Thanks for linking the other issue, I couldn't find it.

**Comment by kchibisov** [comment]
Ideally your system shouldn't enable AA for bitmap fonts, that's how it is done with fontconfig on Linux/BSD for example. However I wonder whether it could be done automatically on macOS. For now I see that we're forcing AA, and knowing that iterm doesn't do it dynamically I feel like there's no way to get it.

**Comment by jeremija** [comment]
Thanks for your reply! Not sure if it helps, but I believe this particular font is a vector font, but the old bitmap version is still embedded within the same font.

From [Wikipedia](https://en.wikipedia.org/wiki/Monaco_(typeface)):

> Monaco has been released in at least three forms. The original was a bitmap monospace font that still appears in the ROMs of even New World Macs, and is still available in recent macOS releases (size 9, with disabled antialiasing).

I am also not able to use other bitmap fonts like Terminus on macOS. Here's an example (left is Alacritty, right is iTerm):

![image](https://user-images.githubusercontent.com/1489493/76688648-103ab380-662f-11ea-8a85-cebe5c3a6a3d.png)

Is this because of what you mentioned above, that AA is being forced?

**Comment by kchibisov** [comment]
>Is this because of what you mentioned above, that AA is being forced?

seems like it, unfortunately, I'm not familiar with macOS font code in alacritty and don't have macOS machine to explore whether it's possible to get AA information from font.

**Comment by jeremija** [comment]
Got it, thanks for all the information!

**Comment by jeremija** [comment]
For what it's worth, I was able to get Terminus to render without antialiasing by setting [all of these](https://github.com/alacritty/alacritty/blob/e2725721686627c976f16d513bb12fe7a303c587/font/src/darwin/mod.rs#L504-L511) variables to false.

![image](https://user-images.githubusercontent.com/1489493/76691681-b1386700-664d-11ea-8731-ad1bdfbd89dc.png)

Monaco 9 looks almost good, however the letter `r` looks like a caret for some reason:

![image](https://user-images.githubusercontent.com/1489493/76691693-cc0adb80-664d-11ea-91f9-9a6258005cc8.png)

**Comment by kchibisov** [comment]
I know, but as you can see there're still some issues with this approach, also AA should be applied to scalable fonts, as well as subpixel AA, etc. If you're interested in working on it I'd recommend to understand how folks are rendering bitmap fonts on macOS in general with core text.

**Comment by kchibisov** [comment]
I've reopened issue, since I think we should have good defaults for macOS bitmap fonts rendering, though it's not something common on this platform. I'd just recommend for someone interesting to understand how options mentioned are working, and investigate whether we should now whether font is scalable or not. Ideally we shouldn't apply AA for non scalable fonts. So this issue is just a task of reading apple docs and trying different techniques.

**Comment by jeremija** [comment]
Thanks again for all the info, and for reopening! What should happen in case of fonts like Monaco which are scalable, but also contain an embedded bitmap inside the same font file?

**Comment by kchibisov** [comment]
There're should be a way to know it. For example with fontconfig you have embedded_bitmap bits and you likely can understand what is being used. I'd assume you have something like that on macOS, though I'd recommend just testing out things, etc. Ideally you shouldn't apply AA when embedded bitmap is used, but use AA when use just normal scalable face.

**Comment by jeremija** [comment]
But if the font is set to `Monaco 9`, users still have to decide whether or not to use this font as a bitmap or not. How could the terminal know whether to use the bitmap, or the scalable part of the font for a specific size wihout a specific config setting?

For example, even the default macOS Terminal app has the antialiasing checkbox:

![image](https://user-images.githubusercontent.com/1489493/76691952-8c91be80-6650-11ea-9832-45ab24bf6efd.png)

In Linux there is fontconfig, so font-specific antialiasing can be enabled or disabled system-wide, but there is no way to configure fonts like that in macOS, as far as I know. I think that's why all these apps usually have a checkbox :)

**Comment by kchibisov** [comment]
>In Linux there is fontconfig, so font-specific antialiasing can be enabled or disabled system-wide, but there is no way to configure fonts like that in macOS, as far as I know. I think that's why all these apps usually have a checkbox :)

But how do they work in a mixed setups, like you have a primary font bitmap, but some fonts are scalable, let's just CJK characters. For example, how are you using bitmaps in `Monaco`?

**Comment by jeremija** [comment]
This is going to be a long post. I'd just like to share my findings in case anyone else runs into this issue.

I am not an expert on fonts, but I've always assumed that the OS provides the substitute for missing characters in bitmap fonts (at least on macOS), because I could always still see macOS emojis in terminal, no matter which font was selected.

I've experimented a bit more and found the following:

# Detect bitmap-only fonts

Bitmap-only fonts can be detected using `font_format` function in  `CTFontDescriptor` and comparing to `core_text::font_descriptor::kCTFontFormatBitmap`.

This method will work with [`GohuFont`](https://font.gohu.org/) (I tested it with [`GohuFont.dfont`](https://font.gohu.org/gohufont.dfont)).

However, the `Monaco` font will have a value of `kCTFontFormatTrueType`.

I made a small utility to print macOS font information using `core-text` [jeremija/rust-core-font-example](https://github.com/jeremija/rust-core-font-example):

```
$ cargo run Monaco GohuFont
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/core-font-test Monaco GohuFont`
Descriptors for font: 'Monaco'
--
name: 'Monaco'
display name: 'Monaco'
is_bitmap: false
is_truetype: true
bold: false
Descriptors for font: 'GohuFont'
--
name: 'GohuFontMedium'
display name: 'GohuFont-Medium'
is_bitmap: true
is_truetype: false
bold: false
--
name: 'GohuFontBold'
display name: 'GohuFont-Bold'
is_bitmap: true
is_truetype: false
bold: true
```

# Alacritty correctly renders bitmap-only fonts

Source code modifications are not necessary when using a bitmap-only font. I assume the problem with Monaco is that it has both vector and bitmap data in the same font. Additionally, Alacritty still correctly renders special characters such as CJK. See the examples below using `GohuFont`, size 11:

Example with all antialiasing/smoothing flags turned on (unmodified source code):

![image](https://user-images.githubusercontent.com/1489493/76700877-2f335700-66bc-11ea-91dc-7a2ddf54861f.png)

Example with all antialiasing/smoothing flags turned off ([all flags in the link set to false](https://github.com/alacritty/alacritty/blob/e2725721686627c976f16d513bb12fe7a303c587/font/src/darwin/mod.rs#L504-L511)):

![image](https://user-images.githubusercontent.com/1489493/76700884-3eb2a000-66bc-11ea-92d2-b70f62827cdf.png)

Note how the chinese characters are antialiased in the first example, so I guess a fallback font is used there? As I mentioned in the beginning, I am not sure if this is handled in alacritty, or on the os-level.

Adding the source of the test file here just in case:

```
U+2EB3 ⺳ CJK RADICAL NET TWO
U+2EB4 ⺴ CJK RADICAL NET THREE
U+2EB5 ⺵ CJK RADICAL NET FOUR

U+1F600 😀 grinning face
U+1F603 😃 grinning face with big eyes
```

# Workaround for Monaco

I managed to convert Monaco to a bitmap-only font using [`FontForge`](https://fontforge.org/). I would have shared the resulting font, but I am not sure if the font license allows itThis can probably be scripted, but I'm not that familiar with the FontForge's scripting API so I'm just going to explain the steps:

1. Open `~/System/Library/Fonts/Monaco.dfont` in `FontForge`.
2. Element > Font Info
3. Change all fields where `Monaco` is mentioned to `MonacoBitmap`
4. Click OK and let FontForge regenerate the font's Unique ID (XUID).
5. File > Generate Font
6. Select `No Outline Font` and `Apple bitmap only sfnt (dfont)`. Save to `~/Library/Fonts/MonacoBitmap.dfont`.

After verifying that the font is installed with `fc-list | grep MonacoBitmap`, set font to `MonacoBitmap` and set size to `9`, and the h and v-offset to `1` in `alacritty.yml`.

Optinally, verify that the font is recognzied as bitmap using my tool [jeremija/rust-core-font-example](https://github.com/jeremija/rust-core-font-example):

```
$ cargo run 'MonacoBitmap'
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/core-font-test MonacoBitmap`
Descriptors for font: 'MonacoBitmap'
--
name: 'MonacoBitmap'
display name: 'MonacoBitmap'
is_bitmap: true
is_truetype: false
bold: false
```

This is the result using Alacritty with no modifications:

![image](https://user-images.githubusercontent.com/1489493/76700927-b1238000-66bc-11ea-942e-6cb9425d81c8.png)

The only character not rendered correctly in Alacritty when using this font is the `~` (tilde) character. I'll have to see if this can be fixed.

As an added bonus, the font stays bitmap on the retina screen now!

**Comment by jeremija** [comment]
I am not sure if this is as bug in the Alacritty, or if the boundaries in MonacoBitmap font are incorrectly defined. A few other characters other than `~` were incorrectly rendered. For example `$` and `Ž`.

![image](https://user-images.githubusercontent.com/1489493/76703470-2e0d2480-66d2-11ea-916c-1573c3e3df1b.png)

I was able to get them to render correctly by patching the following file:

```diff
diff --git a/font/src/darwin/mod.rs b/font/src/darwin/mod.rs
index b9266a8..0e29737 100644
--- a/font/src/darwin/mod.rs
+++ b/font/src/darwin/mod.rs
@@ -392,7 +392,7 @@ impl Font {

         Rect::new(
             Point2D::new(cg_rect.origin.x, cg_rect.origin.y),
-            Size2D::new(cg_rect.size.width, cg_rect.size.height),
+            Size2D::new(cg_rect.size.width + 1., cg_rect.size.height + 3.),
         )
     }

```

![image](https://user-images.githubusercontent.com/1489493/76703460-22216280-66d2-11ea-94fb-5e9409e8fe5e.png)
